File: backend/app/api/v1/spiders.py
"""爬虫任务接口 —— API 层只做参数校验与 Service 编排，不碰 ORM/Session

响应格式：直接返回 domain-specific Pydantic 模型（SpiderTaskResponse 等），
不使用 ApiResponse 包装（与 auth/admin 模块的 ok() 包装不同）。
"""
from typing import Optional

from datetime import datetime

from fastapi import APIRouter, Depends, Path, Query
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.api._helpers import record_audit
from backend.app.api.deps import CurrentUser, require_admin, require_login, require_operator
from backend.services.alert_service import AlertService
from backend.services.schedule_service import ScheduleService
from backend.services.spider_service import SpiderService
from platform_core.db import get_async_db
from platform_core.schemas.spider import (
    AlertRuleRequest,
    AlertRuleResponse,
    AlertRuleUpdateRequest,
    DefinitionCreateRequest,
    DefinitionUpdateMetaRequest,
    DefinitionUpdateRequest,
    RunSpiderRequest,
    ScheduleRequest,
    ScheduleUpdateRequest,
    SpiderDefinitionResponse,
    SpiderFileListResponse,
    SpiderRegistryResponse,
    SpiderResultListResponse,
    SpiderScheduleListResponse,
    SpiderScheduleResponse,
    SpiderTaskListResponse,
    SpiderTaskResponse,
    TaskControlRequest,
    TaskLogResponse,
    TaskQualityReportResponse,
    TaskStoreStatusResponse,
    TaskTemplateRequest,
    TaskTemplateResponse,
    TaskTemplateUpdateRequest,
    TaskUpdateRequest,
    WorkerNodeListResponse,
)

router = APIRouter()


def _service(session: AsyncSession = Depends(get_async_db)) -> SpiderService:
    return SpiderService(session)


def _schedule_service(session: AsyncSession = Depends(get_async_db)) -> ScheduleService:
    return ScheduleService(session)


@router.get("/tasks", response_model=SpiderTaskListResponse)
async def list_tasks(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    status: Optional[str] = Query(None, description="pending/running/completed/failed"),
    priority: Optional[str] = Query(None, pattern="^(high|normal|low)$", description="按优先级过滤"),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderTaskListResponse:
    """获取爬虫任务列表（支持分页和状态/优先级筛选）"""
    return await service.list_tasks(skip=skip, limit=limit, status=status, priority=priority)


@router.post("/run", response_model=SpiderTaskResponse)
async def run_spider(
    payload: RunSpiderRequest,
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> SpiderTaskResponse:
    """入队一次爬虫任务（params 为 JSON 字符串，如 '{"urls": ["https://..."]}'；可指定优先级）"""
    task = await service.enqueue(
        spider_name=payload.spider_name, params=payload.params, priority=payload.priority
    )
    await record_audit(session, user, "task.run", f"task#{task.id}",
                 {"spider": payload.spider_name, "priority": payload.priority})
    return task


@router.get("/results", response_model=SpiderResultListResponse)
async def search_results(
    spider_name: Optional[str] = Query(None, max_length=100, description="按爬虫名过滤（不传查全部）"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    start_time: Optional[datetime] = Query(None, description="采集时间起（ISO 8601）"),
    end_time: Optional[datetime] = Query(None, description="采集时间止（ISO 8601）"),
    keyword: Optional[str] = Query(None, max_length=100, description="关键词（匹配标题/URL/内容）"),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderResultListResponse:
    """跨任务分页查询采集结果（数据中心：爬虫/时间范围/关键词过滤）"""
    return await service.search_results(
        spider_name=spider_name,
        page=page,
        page_size=page_size,
        start_time=start_time,
        end_time=end_time,
        keyword=keyword,
    )


@router.delete("/results/{result_id}")
async def delete_result(
    result_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> dict:
    """删除单条采集结果（数据中心清理；仅管理员）"""
    result = await service.delete_result(result_id)
    await record_audit(session, user, "result.delete", f"result#{result_id}")
    return result


@router.get("/results/{task_id}", response_model=SpiderResultListResponse)
async def list_results(
    task_id: int = Path(..., ge=1),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderResultListResponse:
    """查询指定任务的采集结果（数据闭环出口）"""
    return await service.list_results(task_id=task_id, skip=skip, limit=limit)


@router.get("/results/{task_id}/export")
async def export_results(
    task_id: int = Path(..., ge=1),
    format: str = Query("csv", pattern="^(csv|json)$", description="导出格式：csv/json"),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> Response:
    """导出指定任务的全部采集结果（下载附件）"""
    content, filename, media_type = await service.export_results(task_id, format)
    return Response(
        content=content,
        media_type=media_type,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get("/tasks/{task_id}/store", response_model=TaskStoreStatusResponse)
async def task_store_status(
    task_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> TaskStoreStatusResponse:
    """任务的额外存储目标状态（4.2：目标清单 / redis 缓存条数 / csv 落盘路径）"""
    return await service.store_status(task_id)


@router.get("/registry", response_model=SpiderRegistryResponse)
async def get_registry(
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderRegistryResponse:
    """爬虫注册表：类型表单定义 + 可调度爬虫清单（新增任务弹窗的数据源；清单 DB 优先）"""
    return await service.registry()


@router.get("/nodes", response_model=WorkerNodeListResponse)
async def list_nodes(
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> WorkerNodeListResponse:
    """Worker 节点列表（心跳在线状态 + 各爬虫活跃任务，数据源为 Redis 心跳键）"""
    return await service.list_nodes()


@router.get("/files", response_model=SpiderFileListResponse)
async def list_spider_files(
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderFileListResponse:
    """代码爬虫文件清单（4.4：只读文件元数据 + 关联启停状态）"""
    return await service.spider_files()


@router.patch("/definitions/{name}", response_model=SpiderDefinitionResponse)
async def update_definition(
    payload: DefinitionUpdateRequest,
    name: str = Path(..., min_length=1, max_length=50),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> SpiderDefinitionResponse:
    """启停代码爬虫（4.4：写 spider_definitions.enabled，仅 admin）"""
    definition = await service.update_definition(name, payload.enabled)
    await record_audit(session, user, "definition.update", f"definition#{name}",
                 {"enabled": payload.enabled})
    return definition


@router.post("/definitions", response_model=SpiderDefinitionResponse)
async def create_definition(
    payload: DefinitionCreateRequest,
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> SpiderDefinitionResponse:
    """新建爬虫定义（手动登记，来源标记 manual；仅管理员）"""
    definition = await service.create_definition(payload)
    await record_audit(session, user, "definition.create", f"definition#{payload.name}",
                 {"type": payload.type, "source": "manual"})
    return definition


@router.patch("/definitions/{name}/meta", response_model=SpiderDefinitionResponse)
async def update_definition_meta(
    payload: DefinitionUpdateMetaRequest,
    name: str = Path(..., min_length=1, max_length=50),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> SpiderDefinitionResponse:
    """编辑爬虫定义元信息（标题/描述；仅管理员）"""
    definition = await service.update_definition_meta(name, payload)
    await record_audit(session, user, "definition.update_meta", f"definition#{name}",
                 payload.model_dump(exclude_unset=True))
    return definition


@router.delete("/definitions/{name}")
async def delete_definition(
    name: str = Path(..., min_length=1, max_length=50),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> dict:
    """删除爬虫定义（存在历史任务引用时拒绝；仅管理员）"""
    result = await service.delete_definition(name)
    await record_audit(session, user, "definition.delete", f"definition#{name}")
    return result


@router.patch("/tasks/{task_id}", response_model=SpiderTaskResponse)
async def update_task(
    payload: TaskUpdateRequest,
    task_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> SpiderTaskResponse:
    """编辑待执行任务（仅 pending/queued 可改参数/优先级；待执行任务改优先级会同步搬迁队列）"""
    task = await service.update_task(
        task_id, params=payload.params, priority=payload.priority
    )
    await record_audit(session, user, "task.update", f"task#{task_id}",
                 payload.model_dump(exclude_unset=True))
    return task


@router.delete("/tasks/{task_id}")
async def delete_task(
    task_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> dict:
    """删除任务及其采集结果（running 状态拒绝删除；仅管理员）"""
    result = await service.delete_task(task_id)
    await record_audit(session, user, "task.delete", f"task#{task_id}")
    return result


@router.post("/tasks/{task_id}/control")
async def control_task(
    payload: TaskControlRequest,
    task_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> dict:
    """控制运行中的任务：暂停/恢复/终止（A4；body 必填，缺失时 422）"""
    result = await service.control_task(task_id, payload.action)
    await record_audit(session, user, "task.control", f"task#{task_id}", {"action": payload.action})
    return result


@router.get("/tasks/{task_id}/logs", response_model=TaskLogResponse)
async def get_task_logs(
    task_id: int = Path(..., ge=1),
    lines: int = Query(200, ge=1, le=500),
    keyword: Optional[str] = Query(None, description="全文搜索关键词（大小写不敏感）"),
    level: Optional[str] = Query(None, description="日志级别过滤：DEBUG/INFO/WARNING/ERROR/CRITICAL"),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> TaskLogResponse:
    """任务运行日志（尾部 N 行，支持关键词搜索和级别过滤）"""
    return await service.task_logs(task_id, lines=lines, keyword=keyword, level=level)


@router.get("/tasks/{task_id}/quality", response_model=TaskQualityReportResponse)
async def get_task_quality(
    task_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> TaskQualityReportResponse:
    """获取任务数据质量报告（B1：平均/最低/最高分 + 四档分布）"""
    return await service.get_task_quality(task_id)


# ----------------------------------------------------------------------
# 定时调度（对标 Crawlab 定时任务）
# ----------------------------------------------------------------------
@router.get("/schedules", response_model=SpiderScheduleListResponse)
async def list_schedules(
    service: ScheduleService = Depends(_schedule_service),
    _user: CurrentUser = Depends(require_login),
) -> SpiderScheduleListResponse:
    """调度计划列表"""
    return await service.list_schedules()


@router.post("/schedules", response_model=SpiderScheduleResponse)
async def create_schedule(
    payload: ScheduleRequest,
    service: ScheduleService = Depends(_schedule_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> SpiderScheduleResponse:
    """创建调度计划（校验爬虫注册表 / cron 合法性 / 同爬虫唯一；仅管理员）"""
    schedule = await service.create_schedule(payload)
    await record_audit(session, user, "schedule.create", payload.spider_name,
                 {"cron": payload.cron_expr, "enabled": payload.enabled})
    return schedule


@router.patch("/schedules/{schedule_id}", response_model=SpiderScheduleResponse)
async def update_schedule(
    payload: ScheduleUpdateRequest,
    schedule_id: int = Path(..., ge=1),
    service: ScheduleService = Depends(_schedule_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> SpiderScheduleResponse:
    """更新调度计划（启停 / 改表达式 / 改参数；仅管理员）"""
    schedule = await service.update_schedule(schedule_id, payload)
    await record_audit(session, user, "schedule.update", f"schedule#{schedule_id}",
                 payload.model_dump(exclude_unset=True))
    return schedule


@router.delete("/schedules/{schedule_id}")
async def delete_schedule(
    schedule_id: int = Path(..., ge=1),
    service: ScheduleService = Depends(_schedule_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_admin),
) -> dict:
    """删除调度计划（仅管理员）"""
    result = await service.delete_schedule(schedule_id)
    await record_audit(session, user, "schedule.delete", f"schedule#{schedule_id}")
    return result


# ----------------------------------------------------------------------
# 代理池健康管理（B3）
# ----------------------------------------------------------------------
@router.get("/proxy-health")
async def get_proxy_health(
    _user: CurrentUser = Depends(require_operator),
) -> list[dict]:
    """代理评分排行（评分驱动的智能代理管理）"""
    from backend.services.proxy_health_service import ProxyHealthService

    service = ProxyHealthService()
    return await service.get_proxy_health()


# ----------------------------------------------------------------------
# 告警规则管理（B2）
# ----------------------------------------------------------------------
@router.get("/alert-rules")
async def list_alert_rules(
    session: AsyncSession = Depends(get_async_db),
    _user: CurrentUser = Depends(require_operator),
) -> list[dict]:
    """获取告警规则列表"""
    svc = AlertService(session)
    return await svc.list_rules()


@router.post("/alert-rules", response_model=AlertRuleResponse)
async def create_alert_rule(
    body: AlertRuleRequest,
    session: AsyncSession = Depends(get_async_db),
    _user: CurrentUser = Depends(require_admin),
) -> AlertRuleResponse:
    """创建告警规则（仅管理员）"""
    svc = AlertService(session)
    result = await svc.create_rule(body.model_dump())
    return AlertRuleResponse(**result)


@router.patch("/alert-rules/{rule_id}", response_model=AlertRuleResponse)
async def update_alert_rule(
    rule_id: int = Path(..., ge=1),
    body: AlertRuleUpdateRequest = ...,
    session: AsyncSession = Depends(get_async_db),
    _user: CurrentUser = Depends(require_admin),
) -> AlertRuleResponse:
    """更新告警规则（仅管理员）"""
    svc = AlertService(session)
    result = await svc.update_rule(rule_id, body.model_dump(exclude_unset=True))
    return AlertRuleResponse(**result)


@router.delete("/alert-rules/{rule_id}")
async def delete_alert_rule(
    rule_id: int = Path(..., ge=1),
    session: AsyncSession = Depends(get_async_db),
    _user: CurrentUser = Depends(require_admin),
) -> dict:
    """删除告警规则（仅管理员）"""
    svc = AlertService(session)
    return await svc.delete_rule(rule_id)


# ----------------------------------------------------------------------
# 任务模板（C1）：收藏常用任务配置，一键创建任务
# ----------------------------------------------------------------------
@router.get("/templates", response_model=list[TaskTemplateResponse])
async def list_templates(
    service: SpiderService = Depends(_service),
    _user: CurrentUser = Depends(require_login),
) -> list[TaskTemplateResponse]:
    """获取所有任务模板"""
    return await service.list_templates()


@router.post("/templates", response_model=TaskTemplateResponse)
async def create_template(
    payload: TaskTemplateRequest,
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> TaskTemplateResponse:
    """创建任务模板（收藏当前任务配置）"""
    template = await service.create_template(
        payload.model_dump(), created_by=user.id
    )
    await record_audit(session, user, "template.create", f"template#{template.id}",
                 {"name": payload.name, "spider": payload.spider_name})
    return template


@router.patch("/templates/{template_id}", response_model=TaskTemplateResponse)
async def update_template(
    payload: TaskTemplateUpdateRequest,
    template_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> TaskTemplateResponse:
    """更新任务模板"""
    template = await service.update_template(
        template_id, payload.model_dump(exclude_unset=True)
    )
    await record_audit(session, user, "template.update", f"template#{template_id}")
    return template


@router.delete("/templates/{template_id}")
async def delete_template(
    template_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> dict:
    """删除任务模板"""
    result = await service.delete_template(template_id)
    await record_audit(session, user, "template.delete", f"template#{template_id}")
    return result


@router.post("/templates/{template_id}/run", response_model=SpiderTaskResponse)
async def run_from_template(
    template_id: int = Path(..., ge=1),
    service: SpiderService = Depends(_service),
    session: AsyncSession = Depends(get_async_db),
    user: CurrentUser = Depends(require_operator),
) -> SpiderTaskResponse:
    """从模板创建并运行任务"""
    task = await service.create_task_from_template(template_id)
    await record_audit(session, user, "task.run_from_template", f"task#{task.id}",
                 {"template_id": template_id})
    return task

File: frontend/admin/src/services/spiders.ts
/**
 * 爬虫模块服务 - 任务/注册表/日志/删除 API 封装
 *
 * 响应为后端统一信封（ADR-001）：分页端点 tasks/results 为 PaginatedResponse
 * （data.items/total），其余为 ApiResponse；service 层统一解包 data，
 * 页面组件拿到的仍是裸结构。例外：exportResults 为二进制流下载（白名单，不解包）。
 */
import api, { unwrap } from './api'

export interface SpiderParamField {
  name: string
  label: string
  kind: 'urls' | 'text' | 'json' | 'select' | string
  required?: boolean
  default?: string | null
  help?: string | null
  options?: { value: string; label: string }[] | null
}

export interface SpiderTypeInfo {
  type: string
  label: string
  fields: SpiderParamField[]
}

export interface SpiderInfo {
  name: string
  title: string
  type: string
  description?: string
}

export interface SpiderRegistry {
  types: SpiderTypeInfo[]
  spiders: SpiderInfo[]
}

export interface Task {
  id: number
  spider_name: string
  status: 'pending' | 'running' | 'completed' | 'failed' | string
  priority?: 'high' | 'normal' | 'low' | string
  result_count: number
  retry_count?: number
  error_message?: string | null
  params?: string | null
  created_at?: string | null
  started_at?: string | null
  completed_at?: string | null
}

export interface TaskLogResponse {
  task_id: number
  spider_name: string
  status: string
  lines: string[]
}

export interface SpiderResult {
  id: number
  task_id: number
  spider_name: string
  url?: string | null
  title?: string | null
  content?: string | null
  source?: string | null
  item_type?: string | null
  extra?: string | null
  content_hash?: string | null
  created_at?: string | null
}

export interface SpiderSchedule {
  id: number
  spider_name: string
  cron_expr: string
  params?: string | null
  enabled: boolean
  last_run_at?: string | null
  next_run_at?: string | null
  created_at?: string | null
}

export interface TaskStoreStatus {
  task_id: number
  targets: string[]
  redis_count?: number | null
  csv_path?: string | null
}

export interface SpiderFile {
  name: string
  file: string
  size_bytes: number
  registered: boolean
  enabled?: boolean | null
  title?: string | null
}

/** 爬虫注册表（新增任务弹窗的数据源；后端 SPIDER_TYPES 未登记 flow 类型时前端兖底补齐） */
export const fetchRegistry = (): Promise<SpiderRegistry> =>
  api
    .get('/spiders/registry')
    .then((res) => unwrap<SpiderRegistry>(res))
    .then((reg) => {
      if (reg.types && !reg.types.some((t) => t.type === 'flow')) {
        reg.types = [...reg.types, FLOW_TYPE_FALLBACK]
      }
      return reg
    })

/** flow 类型表单定义（与 plan_json 的 flow 契约对齐：selectors/pagination/detail/filters） */
export const FLOW_TYPE_FALLBACK: SpiderTypeInfo = {
  type: 'flow',
  label: '流程化采集',
  fields: [
    { name: 'urls', label: '页面地址', kind: 'urls', required: true, help: '待采集的列表页 URL，可填多个（每行一个）' },
    { name: 'selectors', label: '提取规则', kind: 'selectors', required: true, help: '字段名 + 选择器类型（xpath/css/regex）+ 表达式，可增删多行' },
    { name: 'pagination', label: '分页设置', kind: 'pagination', help: '可选：下一页选择器 + 最大页数，自动翻页' },
    { name: 'detail', label: '详情页设置', kind: 'detail', help: '可选：列表项选择器 + 链接选择器 + 详情页字段规则，二次采集详情页' },
    { name: 'filters', label: '条件过滤', kind: 'filters', help: '可选：按 包含/等于/正则 过滤提取字段值（字段/操作符/值）' },
  ],
}

/** 任务列表（分页信封 data：{items, total, page, page_size, total_pages}，可选优先级筛选） */
export const fetchTasks = (
  skip = 0,
  limit = 20,
  filters?: { priority?: string; status?: string; spider_name?: string }
): Promise<{ total: number; items: Task[] }> =>
  api
    .get('/spiders/tasks', {
      params: { skip, limit, ...filters },
    })
    .then((res) => unwrap<{ total: number; items: Task[] }>(res))

/** 提交新任务（params 为 JSON 字符串，priority 可选：high/normal/low，默认 normal） */
export const runSpider = (
  spider_name: string,
  params: string,
  priority: 'high' | 'normal' | 'low' = 'normal'
): Promise<Task> =>
  api
    .post('/spiders/run', { spider_name, params, priority })
    .then((res) => unwrap<Task>(res))

/** 删除任务（级联删除采集结果） */
export const deleteTask = (taskId: number): Promise<{ task_id: number; removed_results: number }> =>
  api
    .delete(`/spiders/tasks/${taskId}`)
    .then((res) => unwrap<{ task_id: number; removed_results: number }>(res))

/** 控制运行中的任务：暂停/恢复/终止（A4） */
export const controlTask = (
  taskId: number,
  action: 'pause' | 'resume' | 'stop'
): Promise<{ task_id: number; action: string; message: string }> =>
  api
    .post(`/spiders/tasks/${taskId}/control`, { action })
    .then((res) => unwrap<{ task_id: number; action: string; message: string }>(res))

/** 任务运行日志（尾部 N 行，支持关键词搜索和级别过滤） */
export const fetchTaskLogs = (
  taskId: number,
  lines = 200,
  keyword?: string,
  level?: string,
): Promise<TaskLogResponse> => {
  const params: Record<string, unknown> = { lines }
  if (keyword) params.keyword = keyword
  if (level) params.level = level
  return api
    .get(`/spiders/tasks/${taskId}/logs`, { params })
    .then((res) => unwrap<TaskLogResponse>(res))
}

/** 任务采集结果（分页信封 data） */
export const fetchResults = (
  taskId: number,
  skip = 0,
  limit = 50
): Promise<{ total: number; items: SpiderResult[] }> =>
  api
    .get(`/spiders/results/${taskId}`, { params: { skip, limit } })
    .then((res) => unwrap<{ total: number; items: SpiderResult[] }>(res))

/** 任务额外存储目标状态（4.2：目标清单 / redis 缓存条数 / csv 落盘路径） */
export const fetchTaskStore = (taskId: number): Promise<TaskStoreStatus> =>
  api.get(`/spiders/tasks/${taskId}/store`).then((res) => unwrap<TaskStoreStatus>(res))

/** 代码爬虫文件清单（4.4：只读元数据 + 启停状态；data={total, items}） */
export const fetchSpiderFiles = (): Promise<{ total: number; items: SpiderFile[] }> =>
  api.get('/spiders/files').then((res) => unwrap<{ total: number; items: SpiderFile[] }>(res))

/** 启停代码爬虫（4.4：写 spider_definitions.enabled，后端仅 admin） */
export const updateDefinition = (name: string, enabled: boolean): Promise<SpiderFile> =>
  api.patch(`/spiders/definitions/${name}`, { enabled }).then((res) => unwrap<SpiderFile>(res))

/** 结果导出（blob 下载，自动携带鉴权 Token；二进制流白名单，不解信封） */
export const exportResults = async (taskId: number, format: 'csv' | 'json'): Promise<Blob> => {
  const res = await api.get(`/spiders/results/${taskId}/export`, {
    params: { format },
    responseType: 'blob',
  })
  return res as unknown as Blob
}

// ---------------- 待执行任务编辑（阶段一）----------------
/** 编辑待执行任务（仅 pending/queued 可改 params/priority；改优先级会同步搬迁队列） */
export const updateTask = (
  taskId: number,
  payload: { params?: string; priority?: 'high' | 'normal' | 'low' }
): Promise<Task> =>
  api.patch(`/spiders/tasks/${taskId}`, payload).then((res) => unwrap<Task>(res))

// ---------------- 跨任务结果检索（数据中心）----------------
export interface SearchResultQuery {
  spider_name?: string
  keyword?: string
  start_time?: string
  end_time?: string
  page?: number
  page_size?: number
}

/** 跨任务分页查询采集结果（数据中心：爬虫/时间范围/关键词过滤） */
export const searchResults = (
  query: SearchResultQuery = {}
): Promise<{ total: number; items: SpiderResult[] }> => {
  const params: Record<string, unknown> = {
    page: query.page || 1,
    page_size: query.page_size || 20,
  }
  if (query.spider_name) params.spider_name = query.spider_name
  if (query.keyword) params.keyword = query.keyword
  if (query.start_time) params.start_time = query.start_time
  if (query.end_time) params.end_time = query.end_time
  return api
    .get('/spiders/results', { params })
    .then((res) => unwrap<{ total: number; items: SpiderResult[] }>(res))
}

/** 删除单条采集结果（数据中心清理；仅管理员） */
export const deleteResult = (
  resultId: number
): Promise<{ result_id: number; deleted: boolean }> =>
  api
    .delete(`/spiders/results/${resultId}`)
    .then((res) => unwrap<{ result_id: number; deleted: boolean }>(res))

// ---------------- 爬虫定义完整 CRUD（阶段一）----------------
export interface SpiderDefinition {
  id: number
  name: string
  title: string
  type: string
  description?: string | null
  enabled: boolean
  source?: string
}

/** 新建爬虫定义（手动登记，来源标记 manual；仅管理员；type：api/web/custom/flow） */
export const createDefinition = (payload: {
  name: string
  title: string
  type: string
  description?: string
}): Promise<SpiderDefinition> =>
  api.post('/spiders/definitions', payload).then((res) => unwrap<SpiderDefinition>(res))

/** 编辑爬虫定义元信息（标题/描述；仅管理员） */
export const updateDefinitionMeta = (
  name: string,
  payload: { title?: string; description?: string }
): Promise<SpiderDefinition> =>
  api
    .patch(`/spiders/definitions/${name}/meta`, payload)
    .then((res) => unwrap<SpiderDefinition>(res))

/** 删除爬虫定义（存在历史任务引用时后端拒绝；仅管理员） */
export const deleteDefinition = (
  name: string
): Promise<{ name: string; deleted: boolean }> =>
  api
    .delete(`/spiders/definitions/${name}`)
    .then((res) => unwrap<{ name: string; deleted: boolean }>(res))

// ---------------- 定时调度 ----------------
/** 调度计划列表（data={total, items}，非分页） */
export const fetchSchedules = (): Promise<{ total: number; items: SpiderSchedule[] }> =>
  api.get('/spiders/schedules').then((res) => unwrap<{ total: number; items: SpiderSchedule[] }>(res))

/** 创建调度计划 */
export const createSchedule = (payload: {
  spider_name: string
  cron_expr: string
  params?: string | null
  enabled?: boolean
}): Promise<SpiderSchedule> =>
  api.post('/spiders/schedules', payload).then((res) => unwrap<SpiderSchedule>(res))

/** 更新调度计划（启停/改表达式） */
export const updateSchedule = (
  scheduleId: number,
  payload: { cron_expr?: string; params?: string | null; enabled?: boolean }
): Promise<SpiderSchedule> =>
  api
    .patch(`/spiders/schedules/${scheduleId}`, payload)
    .then((res) => unwrap<SpiderSchedule>(res))

/** 删除调度计划 */
export const deleteSchedule = (
  scheduleId: number
): Promise<{ schedule_id: number; spider_name: string }> =>
  api
    .delete(`/spiders/schedules/${scheduleId}`)
    .then((res) => unwrap<{ schedule_id: number; spider_name: string }>(res))

// ---------------- 告警规则 ----------------
export interface AlertRule {
  id: number
  name: string
  spider_name?: string | null
  rule_type: string
  threshold: number
  window_minutes: number
  severity: string
  channels?: string[] | null
  enabled: boolean
  last_triggered_at?: string | null
  created_at?: string | null
}

/** 告警规则列表（data=[...]） */
export const fetchAlertRules = (): Promise<AlertRule[]> =>
  api.get('/spiders/alert-rules').then((res) => unwrap<AlertRule[]>(res))

/** 创建告警规则 */
export const createAlertRule = (payload: {
  name: string
  spider_name?: string | null
  rule_type: string
  threshold: number
  window_minutes?: number
  severity?: string
  channels?: string[] | null
  enabled?: boolean
}): Promise<AlertRule> =>
  api.post('/spiders/alert-rules', payload).then((res) => unwrap<AlertRule>(res))

/** 更新告警规则 */
export const updateAlertRule = (
  ruleId: number,
  payload: {
    name?: string
    threshold?: number
    window_minutes?: number
    severity?: string
    channels?: string[] | null
    enabled?: boolean
  }
): Promise<AlertRule> =>
  api.patch(`/spiders/alert-rules/${ruleId}`, payload).then((res) => unwrap<AlertRule>(res))

/** 删除告警规则 */
export const deleteAlertRule = (ruleId: number): Promise<{ rule_id: number; deleted: boolean }> =>
  api
    .delete(`/spiders/alert-rules/${ruleId}`)
    .then((res) => unwrap<{ rule_id: number; deleted: boolean }>(res))

// ---------------- 任务模板（C1）----------------
export interface TaskTemplate {
  id: number
  name: string
  spider_name: string
  params?: string | null
  priority: string
  created_by?: number | null
  created_at?: string | null
}

/** 模板列表（data=[...]） */
export const fetchTemplates = (): Promise<TaskTemplate[]> =>
  api.get('/spiders/templates').then((res) => unwrap<TaskTemplate[]>(res))

/** 创建模板 */
export const createTemplate = (payload: {
  name: string
  spider_name: string
  params?: string | null
  priority?: string
}): Promise<TaskTemplate> =>
  api.post('/spiders/templates', payload).then((res) => unwrap<TaskTemplate>(res))

/** 更新模板 */
export const updateTemplate = (
  templateId: number,
  payload: {
    name?: string
    spider_name?: string
    params?: string | null
    priority?: string
  }
): Promise<TaskTemplate> =>
  api
    .patch(`/spiders/templates/${templateId}`, payload)
    .then((res) => unwrap<TaskTemplate>(res))

/** 删除模板 */
export const deleteTemplate = (templateId: number): Promise<{ id: number; deleted: boolean }> =>
  api
    .delete(`/spiders/templates/${templateId}`)
    .then((res) => unwrap<{ id: number; deleted: boolean }>(res))

/** 从模板创建并运行任务 */
export const runFromTemplate = (templateId: number): Promise<Task> =>
  api.post(`/spiders/templates/${templateId}/run`).then((res) => unwrap<Task>(res))
